oozie-to-airflow/mappers/fs_mapper.py

```diff
@@ -0,0 +1,198 @@
+    return command
+
+
+def prepare_chgrp_command(node: Element, params):
+    path = normalize_path(node.attrib[FS_TAG_PATH], params)
+    group = normalize_path(node.attrib[FS_TAG_GROUP], params)
+
+    recursive = node.find(FS_TAG_RECURSIVE) is not None
+    extra_param = "-R" if recursive else ""
+
+    command = "fs -chgrp {extra} {group} {path}".format(
+        extra=extra_param, group=shlex.quote(group), path=shlex.quote(path)
+    )
+    return command
+
+
+def prepare_setrep_command(node: Element, params):
+    path = normalize_path(node.attrib[FS_TAG_PATH], params)
+    fac = node.attrib[FS_TAG_REPLFAC]
+
+    command = "fs -setrep {fac} {path}".format(fac=shlex.quote(fac), path=shlex.quote(path))
+    return command
+
+
+FS_OPERATION_MAPPERS = {
+    FS_OP_MKDIR: prepare_mkdir_command,
+    FS_OP_DELETE: prepare_delete_command,
+    FS_OP_MOVE: prepare_move_command,
+    FS_OP_CHMOD: prepare_chmod_command,
+    FS_OP_TOUCHZ: prepare_touchz_command,
+    FS_OP_CHGRP: prepare_chgrp_command,
+    FS_OP_SETREP: prepare_setrep_command,
+}
+
+
+class Task(NamedTuple):
+    task_id: str
+    rendered_template: str
+
+
+def chain(ops):
+    return [Relation(from_task_id=a.task_id, to_task_id=b.task_id) for a, b in zip(ops, ops[1::])]
+
+
+class FsMapper(ActionMapper):
+    """
+    Converts a FS Oozie node to an Airflow task.
+    """
+
+    tasks: List[Task]
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.tasks = []
+
+    def on_parse_node(self):
+        super().on_parse_node()
+        self.tasks = self.parse_tasks()
+
+    def parse_tasks(self):
+        if not list(self.oozie_node):
+            return [
+                Task(
+                    task_id=self.name,
+                    rendered_template=render_template(
+                        "dummy.tpl", task_id=self.name, trigger_rule=self.trigger_rule
+                    ),
+                )
+            ]
+
+        return [self.parse_fs_action(i, node) for i, node in enumerate(self.oozie_node)]
+
+    def convert_to_text(self):
+        return render_template(
+            template_name="fs.tpl",
+            task_id=self.name,
+            trigger_rule=self.trigger_rule,
+            sub_ops=self.tasks,
+            relations=chain(self.tasks),
+        )
+
+    @staticmethod
+    def required_imports() -> Set[str]:
+        return {
+            "from airflow.operators import dummy_operator",
+            "from airflow.operators import bash_operator",
+            "import shlex",
+        }
+
+    @property
+    def first_task_id(self):
+        return self.tasks[0].task_id
+
+    @property
+    def last_task_id(self):
+        return self.tasks[-1].task_id
+
+    def parse_fs_action(self, index: int, node: Element):
+        tag_name = node.tag
+        tasks_count = len(self.oozie_node)
+        task_id = self.name if tasks_count == 1 else "{}_fs_{}_{}".format(self.name, index, tag_name)
```

Comment:
I believe we agreed on using the `f"` convention instead of `.format()` for formatting strings.

Reply:
I updated a PR.